wire-ios-data-model/Source/Model/Conversation/ZMConversation+Predicates.swift

```diff
@@ -77,102 +77,124 @@ extension ZMConversation {
+        let hasNoUserDefinedName = NSPredicate(format: "\(ZMConversationUserDefinedNameKey) == NULL")
+        let hasOnlyOneParticipant = NSPredicate(format: "\(ZMConversationParticipantRolesKey).@count == 2")
+
+        return NSCompoundPredicate(andPredicateWithSubpredicates: [isTeamConversation, isGroupConversation, hasNoUserDefinedName, hasOnlyOneParticipant])
     }
 
-    @objc(predicateForClearedConversations)
-    class func predicateForClearedConversations() -> NSPredicate {
-        let cleared = NSPredicate(format: "\(ZMConversationClearedTimeStampKey) != NULL AND \(ZMConversationIsArchivedKey) == YES")
+    class func predicateForConversationsNeedingToBeCalculatedUnreadMessages() -> NSPredicate {
+         return NSPredicate(format: "%K == YES", ZMConversationNeedsToCalculateUnreadMessagesKey)
+    }
 
-        return NSCompoundPredicate(andPredicateWithSubpredicates: [cleared, predicateForValidConversations()])
+    public static func predicateForConversationsArePendingToRefreshMetadata() -> NSPredicate {
+        return NSPredicate(format: "\(ZMConversationIsPendingMetadataRefreshKey) == YES")
     }
 
-    @objc(predicateForConversationsIncludingArchived)
-    class func predicateForConversationsIncludingArchived() -> NSPredicate {
-        let notClearedTimestamp = NSPredicate(format: "\(ZMConversationClearedTimeStampKey) == NULL OR \(ZMConversationLastServerTimeStampKey) > \(ZMConversationClearedTimeStampKey) OR (\(ZMConversationLastServerTimeStampKey) == \(ZMConversationClearedTimeStampKey) AND \(ZMConversationIsArchivedKey) == NO)")
+}
 
-        return NSCompoundPredicate(andPredicateWithSubpredicates: [notClearedTimestamp, predicateForValidConversations()])
-    }
+extension String {
 
-    @objc(predicateForGroupConversations)
-    class func predicateForGroupConversations() -> NSPredicate {
-        let groupConversationPredicate = NSPredicate(format: "\(ZMConversationConversationTypeKey) == \(ZMConversationType.group.rawValue)")
-        let notInFolderPredicate = NSCompoundPredicate(notPredicateWithSubpredicate: predicateForConversationsInFolders())
+    var words: [String] {
+        var words: [String] = []
+        enumerateSubstrings(in: self.startIndex..., options: .byWords) { substring, _, _, _ in
+            words.append(String(substring!))
+        }
 
-        return .all(of: [
-            predicateForConversationsExcludingArchived(),
-            groupConversationPredicate,
-            notInFolderPredicate
-        ])
+        if words.isEmpty {
+            words = [trimmingCharacters(in: .whitespacesAndNewlines)]
+        }
+
+        return words
     }
+}
 
-    @objc(predicateForLabeledConversations:)
-    class func predicateForLabeledConversations(_ label: Label) -> NSPredicate {
-        let labelPredicate = NSPredicate(format: "%@ IN \(ZMConversationLabelsKey)", label)
+@objc
+public final class ConversationPredicateFactory: NSObject {
```

Comment:
Please let's move this factory to a dedicated file as it is better searchable and also easier to move to other modules in future.

Reply:
Moved to a new file.